File: ui/src/api/raybot-command.ts
import type { SortPrefix } from '@/lib/sort'
import type { Paging } from '@/types/paging'
import type { InputMap, RaybotCommand, RaybotCommandType } from '@/types/raybot-command'
import http from '@/lib/http'

export type ListRaybotCommandSort = 'status' | 'created_at' | 'completed_at'
export interface ListRaybotCommandParams {
  page?: number
  pageSize?: number
  sort?: SortPrefix<ListRaybotCommandSort>[]
}

export interface CreateRaybotCommandParams<T extends RaybotCommandType> {
  type: T
  input?: InputMap[T]
}

const raybotCommands = {
  list(raybotId: string, p: ListRaybotCommandParams): Promise<Paging<RaybotCommand>> {
    return http.get(`/raybots/${raybotId}/commands`, { params: {
      page: p.page,
      pageSize: p.pageSize,
      sort: p.sort?.join(','),
    },
    })
  },
  get(id: string): Promise<RaybotCommand> {
    return http.get(`/raybot-commands/${id}`)
  },
  create<T extends RaybotCommandType>(
    raybotId: string,
    body: CreateRaybotCommandParams<T>,
  ): Promise<RaybotCommand<T>> {
    return http.post(`/raybots/${raybotId}/commands`, body)
  },
}

export default raybotCommands

File: internal/controller/http/handler/raybot_command.go
package handler

import (
	"context"
	"fmt"

	"github.com/tuanvumaihuynh/roboflow/internal/controller/http/oas/converter"
	"github.com/tuanvumaihuynh/roboflow/internal/controller/http/oas/gen"
	raybotcommand "github.com/tuanvumaihuynh/roboflow/internal/model/raybot_command"
	"github.com/tuanvumaihuynh/roboflow/internal/service"
	"github.com/tuanvumaihuynh/roboflow/pkg/paging"
	"github.com/tuanvumaihuynh/roboflow/pkg/sort"
)

type raybotCommandHandler struct {
	raybotCommandSvc service.RaybotCommandService
}

func newRaybotCommandHandler(raybotCommandSvc service.RaybotCommandService) *raybotCommandHandler {
	return &raybotCommandHandler{raybotCommandSvc: raybotCommandSvc}
}

func (h raybotCommandHandler) RaybotCommandGet(ctx context.Context, request gen.RaybotCommandGetRequestObject) (gen.RaybotCommandGetResponseObject, error) {
	m, err := h.raybotCommandSvc.GetRaybotCommand(ctx, service.GetRaybotCommandParams{
		ID: request.RaybotCommandId,
	})
	if err != nil {
		return nil, err
	}

	return gen.RaybotCommandGet200JSONResponse(converter.ToRaybotCommandResponse(m)), nil
}

func (h raybotCommandHandler) RaybotCommandList(ctx context.Context, request gen.RaybotCommandListRequestObject) (gen.RaybotCommandListResponseObject, error) {
	pagingParams := paging.NewParams(
		request.Params.PageSize,
		request.Params.Page,
		paging.WithMaxPageSize(1000),
	)

	var sorts []sort.Sort
	var err error
	if request.Params.Sort != nil {
		sorts, err = sort.NewListFromString(*request.Params.Sort)
		if err != nil {
			return nil, fmt.Errorf("sort new list from string: %w", err)
		}
	}

	mp, err := h.raybotCommandSvc.ListRaybotCommandsByRaybotID(ctx, service.ListRaybotCommandsByRaybotIDParams{
		RaybotID:     request.RaybotId,
		PagingParams: pagingParams,
		Sorts:        sorts,
	})
	if err != nil {
		return nil, fmt.Errorf("raybot command service list raybot commands by raybot id: %w", err)
	}

	items := make([]gen.RaybotCommandResponse, len(mp.Items))
	for i, item := range mp.Items {
		items[i] = converter.ToRaybotCommandResponse(item)
	}

	return gen.RaybotCommandList200JSONResponse{
		Items:      items,
		TotalItems: mp.TotalItems,
	}, nil
}

func (h raybotCommandHandler) RaybotCommandCreate(ctx context.Context, request gen.RaybotCommandCreateRequestObject) (gen.RaybotCommandCreateResponseObject, error) {
	m, err := h.raybotCommandSvc.CreateRaybotCommand(ctx, service.CreateRaybotCommandParams{
		RaybotID: request.RaybotId,
		Type:     raybotcommand.Type(request.Body.Type),
		Inputs:   raybotcommand.NewInputs(request.Body.Inputs),
	})
	if err != nil {
		return nil, fmt.Errorf("raybot command service create raybot command: %w", err)
	}

	return gen.RaybotCommandCreate201JSONResponse(converter.ToRaybotCommandResponse(m)), nil
}
